packages/teams-js/src/private/externalAppAuthenticationForCEA.ts

```diff
@@ -0,0 +1,269 @@
+        conversationId,
+        authenticateParameters.url.href,
+        authenticateParameters.width,
+        authenticateParameters.height,
+        authenticateParameters.isExternal,
+      ],
+    );
+    if (error) {
+      throw error;
+    }
+  }
+
+  /**
+   * @beta
+   * @hidden
+   * Signals to the host to perform authentication using the given authentication parameters and then resend the request to the application backend with the authentication result.
+   * @internal
+   * Limited to Microsoft-internal use
+   * @param appId ID of the application backend to which the request and authentication response should be sent. This must be a UUID
+   * @param conversationId ConversationId To tell the bot what conversation the calls are coming from
+   * @param authenticateParameters Parameters for the authentication pop-up
+   * @param originalRequestInfo Information about the original request that should be resent
+   * @throws InvokeError if the host encounters an error while authenticating or resending the request
+   * @returns A promise that resolves to the IActionExecuteResponse from the application backend and rejects with InvokeError if the host encounters an error while authenticating or resending the request
+   */
+  export async function authenticateAndResendRequest(
+    appId: AppId,
+    conversationId: string,
+    authenticateParameters: externalAppAuthentication.AuthenticatePopUpParameters,
+    originalRequestInfo: externalAppAuthentication.IActionExecuteInvokeRequest,
+  ): Promise<externalAppAuthentication.IActionExecuteResponse> {
+    ensureInitialized(runtime, FrameContexts.content);
+
+    if (!isSupported()) {
+      throw errorNotSupportedOnPlatform;
+    }
+
+    validateId(conversationId, new Error('conversation id is not valid.'));
+
+    validateOriginalRequestInfo(originalRequestInfo);
+
+    // Ask the parent window to open an authentication window with the parameters provided by the caller.
+    const [response] = await sendMessageToParentAsync<
+      [externalAppAuthentication.InvokeErrorWrapper | externalAppAuthentication.IActionExecuteResponse]
+    >(
+      getApiVersionTag(
+        externalAppAuthenticationTelemetryVersionNumber,
+        ApiName.ExternalAppAuthenticationForCEA_AuthenticateAndResendRequest,
+      ),
+      ApiName.ExternalAppAuthenticationForCEA_AuthenticateAndResendRequest,
+      [
+        appId.toString(),
+        conversationId,
+        originalRequestInfo,
+        authenticateParameters.url.href,
+        authenticateParameters.width,
+        authenticateParameters.height,
+        authenticateParameters.isExternal,
+      ],
+    );
+    if (isActionExecuteResponse(response)) {
+      return response;
+    } else if (isInvokeErrorWrapper(response)) {
+      throw response;
+    } else {
+      throw defaultExternalAppError;
+    }
+  }
+
+  function isActionExecuteResponse(response: unknown): response is externalAppAuthentication.IActionExecuteResponse {
```

Comment:
>isActionExecuteResponse [](http://example.com/codeflow?start=11&length=23)

I usually like to keep type guards defined next to the place where the type itself is defined since they are tightly coupled. Moreover, if anyone needs a `IActionExecuteResponse` type guard in the future, they are less likely to accidentally make a duplicate implementation if the existing one lives right next to the type.

Reply:
Moved them closer to the interfaces